pylon-core/src/main/kotlin/io/github/pylonmc/pylon/core/block/base/PylonCargoBlock.kt

```diff
@@ -0,0 +1,265 @@
+ *
+ * Each face can have one logistic group which cargo ducts connected to that face
+ * are allowed to interface with
+ *
+ * In your place constructor, you will need to call [addCargoLogisticGroup] for all
+ * the block faces you want to be able to connect cargo ducts to, and also
+ * `setCargoTransferRate` to set the maximum number of items that can be transferred
+ * out of this block per cargo tick.
+ */
+interface PylonCargoBlock : PylonLogisticBlock {
+
+    private val cargoBlockData: CargoBlockData
+        get() = cargoBlocks.getOrPut(this) { CargoBlockData(
+            mutableMapOf(),
+            1
+        )}
+
+    @ApiStatus.NonExtendable
+    fun addCargoLogisticGroup(face: BlockFace, group: String) {
+        cargoBlockData.groups.put(face, group)
+    }
+
+    @ApiStatus.NonExtendable
+    fun removeCargoLogisticGroup(face: BlockFace) {
+        cargoBlockData.groups.remove(face)
+    }
+
+    @ApiStatus.NonExtendable
+    fun getCargoLogisticGroup(face: BlockFace): LogisticGroup?
+            = cargoBlockData.groups[face]?.let { getLogisticGroup(it) }
+
+    val cargoLogisticGroups
+        @ApiStatus.NonExtendable
+        get() = cargoBlockData.groups.toMap()
+
+    var cargoTransferRate
+        /**
+         * Note that [cargoTransferRate] will be multiplied by [PylonConfig.cargoTransferRateMultiplier],
+         * and the result will be the maximum number of items that can be transferred
+         * out of this block per cargo tick.
+         *
+         * @see [cargoItemsTransferredPerSecond]
+         */
+        @ApiStatus.NonExtendable
+        set(transferRate) {
+            cargoBlockData.transferRate = transferRate
+        }
+        @ApiStatus.NonExtendable
+        get() = cargoBlockData.transferRate
+
+    fun onDuctConnected(event: PylonCargoDuctConnectEvent) {}
+
+    fun onDuctDisconnected(event: PylonCargoDuctDisconnectEvent) {}
+
+    fun tickCargo() {
+        for ((face, group) in cargoBlockData.groups) {
+            val sourceGroup = getLogisticGroup(group)
+            if (sourceGroup == null || sourceGroup.slotType != LogisticSlotType.OUTPUT) {
+                continue
+            }
+
+            val target = CargoRoutes.getCargoTarget(this, face)
+            if (target == null || target.block.block == block) {
+                continue
+            }
+
+            val targetGroup = target.block.getCargoLogisticGroup(target.face)
+            if (targetGroup == null || targetGroup.slotType != LogisticSlotType.INPUT) {
+                continue
+            }
+
+            tickCargoFace(sourceGroup, targetGroup)
+        }
+    }
+
+    fun tickCargoFace(sourceGroup: LogisticGroup, targetGroup: LogisticGroup) {
+        for (sourceSlot in sourceGroup.slots) {
+            val sourceStack = sourceSlot.getItemStack()
+            val sourceAmount = sourceSlot.getAmount()
+            if (sourceStack == null || (targetGroup.filter != null && !targetGroup.filter!!(sourceStack))) {
```

Comment:
```suggestion
            if (sourceStack == null || (targetGroup.filter?.call(sourceStack) == false)) {
```
I'd write this like this but idrc, feel free to ignore